src/Runtime/OMTensorList.inc

```diff
@@ -26,37 +26,34 @@
 #endif
 
 #include "onnx-mlir/Runtime/OMTensorList.h"
+#include <string.h>
 
 struct OMTensorList {
 #ifdef __cplusplus
-  /**
-   * Constructor
-   *
-   * Create an OMTensorList with specified OMTensor pointer array
-   * and the size of the array
-   */
-  OMTensorList(OMTensor *omts[], int64_t n) : _omts(omts), _size(n) {
-    _owning = false;
-  };
 
   /**
    * Constructor
    *
-   * Create an empty OMTensorList for internal API calls.
+   * Create an OMTensorList with specified OMTensor pointer array, the size of
+   * the array and indication of whether or not to perform a shallow destroy of
+   * the OMTensorList when explicitly calling the OMTensorListDestroyShallow
+   * method
    */
-  OMTensorList() = default;
+  OMTensorList(OMTensor *omts[], int64_t n, bool shallow = false)
+      : _omts(omts), _size(n), _shallow(shallow){};
 
   /**
    * Destructor
    *
    * Destroy the OMTensorList struct.
    */
   ~OMTensorList() {
-    /* Destroy all the OMTensors */
-    for (int64_t i = 0; i < _size; i++)
-      omTensorDestroy(_omts[i]);
-    if (_owning)
-      free(_omts);
+    if (_shallow)
+      omTensorListDestroyShallow(this);
+    else
+      /* Destroy all the OMTensors */
+      omTensorListDestroy(this);
+    free(_omts);
```

Comment:
You are confusing C++ destructor with C destroy function. `this` is a pointer to C++ `OMTensorList` object, it cannot be passed to `omTensorListDestroy` or `omTensorListDestroyShallow` which expects a pointer to C `OMTensorList` struct.

The destructor should be very simple. All you need is to surround the for loop with checking _shallow, i.e.,
```
  ~OMTensorList() {
    /* Destroy all the OMTensors */
    if (!_shallow)
      for (int64_t i = 0; i < _size; i++)
        omTensorDestroy(_omts[i]);
    free(_omts);
  };
```

Reply:
okay